src/watch.rs

```diff
@@ -26,98 +27,122 @@ where
 {
     println!("Watching for changes...");
 
-    let rx_changes = trigger_on_change(&mut watch_sources, &mut watch_code)?;
+    let mut watch_sources_old: HashSet<_> = watch_sources.collect();
+    let mut watch_code_old: HashSet<_> = watch_code.collect();
+
+    trigger_on_change(
+        watch_sources_old.clone(),
+        watch_code_old.clone(),
+        |change, sw, cw| {
+            if allow_reverse || change == ChangeType::Sources {
+                println!(
+                    "{} changed. Re-building...",
+                    if change == ChangeType::Sources {
+                        "Sources"
+                    } else {
+                        "Code"
+                    }
+                );
 
-    for change in rx_changes {
-        if allow_reverse || change == ChangeType::Sources {
-            println!(
-                "{} changed. Re-building...",
-                if change == ChangeType::Sources {
-                    "Sources"
-                } else {
-                    "Code"
-                }
-            );
+                let curr_dir = env::current_dir()?;
+                let (config, mut watch_sources_new, watch_code_new, _has_reverse) =
+                    cmd::run_with_args(&args, Some(change == ChangeType::Code))?;
+                env::set_current_dir(&curr_dir)?;
+
+                watch_sources_new.insert(config);
+
+                update_watcher(sw, &watch_sources_old, &watch_sources_new)?;
+                update_watcher(cw, &watch_code_old, &watch_code_new)?;
+
+                watch_sources_old = watch_sources_new;
+                watch_code_old = watch_code_new;
+            }
+            Ok(())
+        },
+    )?;
 
-            let curr_dir = env::current_dir()?;
-            let (_config, _watch_forward, _watch_reverse, _has_reverse) =
-                cmd::run_with_args(&args, Some(change == ChangeType::Code))?;
-            env::set_current_dir(&curr_dir)?;
+    Ok(())
+}
+
+fn update_watcher(
+    watcher: &mut RecommendedWatcher,
+    old_files: &HashSet<PathBuf>,
+    new_files: &HashSet<PathBuf>,
+) -> Fallible {
+    for path in old_files {
```

Comment:
The [`HashSet::difference`](https://doc.rust-lang.org/stable/std/collections/struct.HashSet.html#method.difference) method is not applicable here?

Reply:
Another thing I was not aware of...